### ISSUE TITLE
`ABC.register` not recognized on `collections.abc.Mapping`

### ISSUE THREAD

**Issue by gerlero** [astral-sh/ty#1204]
### Summary

When using `Mapping.register(...)` to register a virtual subclass, `ty` reports an `unresolved-attribute` error on `register`.

**Code to reproduce:**

```python
from collections.abc import Mapping

class MyClass:
    pass

Mapping.register(MyClass)
```

**Error output:**

```
error[unresolved-attribute]: Type `<class 'Mapping'>` has no attribute `register`
 --> test.py:6:1
  |
4 |     pass
5 |
6 | Mapping.register(MyClass)
  | ^^^^^^^^^^^^^^^^
  |
info: rule `unresolved-attribute` is enabled by default
```
**Notes:**

This suggests to me that `ty` isn’t accounting for the fact that `collections.abc` classes are subclasses of `ABC`.

### Version

ty 0.0.1-alpha.20

**Comment by carljm** [comment]
Thanks for the report!

Pyright and pyrefly both emit the same error as ty here, but mypy does not.

Looking at [the typeshed definition](https://github.com/python/typeshed/blob/main/stdlib/typing.pyi#L764) of `Mapping` vs [the runtime definition](https://github.com/python/cpython/blob/main/Lib/_collections_abc.py#L775), if you follow the inheritance tree, the typeshed definition ends up not inheriting from anything defined with a metaclass of `ABCMeta`, whereas the runtime definition does inherit from `Sized` and `Container`, which have the `ABCMeta` metaclass. (`Container` has the same issue itself -- at runtime it has `ABCMeta`, but not in typeshed.)

For e.g. `Sized` or any other type that actually is defined in typeshed as having the `ABCMeta` metaclass, we [do recognize the `register` method](https://play.ty.dev/f609edb2-3ffe-4c5f-a30d-0631a7206e3a).

So this seems like a typeshed bug; mypy must be doing some kind of special-casing to work around it. Maybe we should, too? Or maybe it should be fixed in typeshed?

This seems like an issue @AlexWaygood might have additional context on?

**Comment by AlexWaygood** [comment]
> So this seems like a typeshed bug; mypy must be doing some kind of special-casing to work around it. Maybe we should, too? Or maybe it should be fixed in typeshed?
> 
> This seems like an issue [@AlexWaygood](https://github.com/AlexWaygood) might have additional context on?

I think the special-casing mypy does is that mypy is aware that all protocols implicitly have `_ProtocolMeta` as their metaclass, and `_ProtocolMeta` is a subclass of `ABCMeta`. This means that any class that has `Protocol` in its MRO has `_ProtocolMeta` as its metaclass, and therefore has a `register` method available on the class:

```pycon
>>> from typing import Protocol
>>> type(Protocol)
<class 'typing._ProtocolMeta'>
>>> _.__mro__
(<class 'typing._ProtocolMeta'>, <class 'abc.ABCMeta'>, <class 'type'>, <class 'object'>)
>>> Protocol.register
<bound method ABCMeta.register of <class 'typing.Protocol'>>
```

We should probably implement similar special casing, since it's certainly true that a protocol class (or a subclass of a protocol class) does not need to explicitly declare `metaclass=ABCMeta` in order for it to have a `register` method available on the class. Once we implement this special casing, we will recognise `Mapping.register()` as existing, since -- according to typeshed -- `Mapping` has `Protocol` in its MRO.

**Comment by carljm** [comment]
Right, so the special-casing comes back to typeshed defining `Protocol` itself as just `Protocol: _SpecialForm`, and giving no details about what `Protocol` actually looks like at runtime. So type-checkers need to replicate via hard-coding all the relevant runtime effects of inheriting `Protocol`. Thanks!

**Comment by AlexWaygood** [comment]
> Right, so the special-casing comes back to typeshed defining `Protocol` itself as just `Protocol: _SpecialForm`, and giving no details about what `Protocol` actually looks like at runtime. So type-checkers need to replicate via hard-coding all the relevant runtime effects of inheriting `Protocol`. Thanks!

Indeed! Which is one of those things that I'd theoretically be very open to changing in typeshed, but which in practice I suspect would be pretty hard to change because of the fact that most type checkers (including us) now have hard-coded logic for recognising `Protocol` with its current typeshed definition :/

**Comment by choru-k** [comment]
Adding a second symptom of this gap, plus a stdlib-only reproduction.

### The metaclass is right for `Protocol` but not for its subclasses

```python
import abc
from typing import Protocol, reveal_type


class P(Protocol):
    def x(self) -> str: ...


class Base(abc.ABC): ...


reveal_type(type(Protocol))  # ty: `<class '_ProtocolMeta'>`  correct
reveal_type(type(P))         # ty: `<class 'type'>`           runtime is _ProtocolMeta
reveal_type(type(Base))      # ty: `<class 'ABCMeta'>`        correct
```

Runtime on 3.12: `('_ProtocolMeta', '_ProtocolMeta', 'ABCMeta')`.

So the special-casing added for #1218 (ruff PR #20488) covers the bare special form; it does not
propagate to a class with `Protocol` in its MRO. That inconsistency is visible in one file, which may
make a convenient test case.

### Second symptom: synthesized metaclasses

Combining a `Protocol` with an ABC hierarchy needs a synthesized metaclass — the conventional fix for
a metaclass conflict. With `type(P)` inferred as `type`, both of ty's diagnostics fire:

```python
import abc
from typing import Protocol


class P(Protocol):
    def x(self) -> str: ...


class Base(abc.ABC): ...


class Meta(type(P), type(Base)):
    pass


class D(Base, P, metaclass=Meta):
    def x(self) -> str:
        return "x"
```

ty 0.0.72, no suppressions:

```text
r.py:12:7: error[inconsistent-mro] Cannot create a consistent method resolution order (MRO) for
  class `Meta` with bases list `[<class 'type'>, <class 'ABCMeta'>]`
r.py:16:1: error[conflicting-metaclass] The metaclass of a derived class (`D`) must be a subclass of
  the metaclasses of all its bases, but `Meta` (metaclass of `D`) and `ABCMeta` (metaclass of base
  class `Base`) have no subclass relationship
```

Note the rendered bases list: `type` where `_ProtocolMeta` belongs. Both diagnostics are correct
*given that premise*. `D().x()` returns `"x"` at runtime.

mypy requires `# type: ignore` on the dynamic-base line (`Unsupported dynamic base class "type"`,
`[misc]`), so that suppression is present in real code anyway. With it, ty still reports line 16.

For context rather than comparison — the other checkers do not model this at all, so the question does
not arise for them: mypy reveals `type[def () -> P]`, pyright `type[type[P]]`. pyrefly needs no
suppression here; it reported the same two-error shape as ty on 1.2.0 and is clean on 1.3.0.dev1 and 1.3.0.dev2
(with a config present — without one it defaults to preset `basic` and reports nothing either way).

### Source

[`crates/ty_python_semantic/src/types/class_base.rs:371`](https://github.com/astral-sh/ruff/blob/82e37beef9b39da7ad66cd27ab94edec8c187eb1/crates/ty_python_semantic/src/types/class_base.rs#L371),
inside `ClassBase::metaclass()`:

```rust
// TODO: all `Protocol` classes actually have `_ProtocolMeta` as their metaclass.
Self::Protocol | Self::Generic | Self::TypedDict(_) => {
    KnownClass::Type.to_instance(db, env)
}
```

`TypedDict` shares the arm and is wrong the same way — runtime metaclass is `_TypedDictMeta`.
`Generic` genuinely is `type`, so that arm is fine.

`mdtest/protocols.md` asserts `reveal_type(type(Protocol))` is `_ProtocolMeta` (with a
`static_assert(is_subtype_of(TypeOf[Protocol], typing._ProtocolMeta))`), but there is no test for the
metaclass of a protocol *subclass*. `mdtest/metaclass.md` has no `Protocol` or `type(...)`-base
coverage.

### On the "true positive given typeshed" reading

[#2052](https://github.com/astral-sh/ty/issues/2052) and [#983](https://github.com/astral-sh/ty/issues/983)
were both closed on the reasoning that the diagnostic is correct given what typeshed declares — #983
specifically because the PySide6/Shiboken stubs never declared the real metaclass. That reasoning does
not transfer here. `_ProtocolMeta` *is* declared in typeshed (`stdlib/typing.pyi:475`, as
`class _ProtocolMeta(ABCMeta)`), but nothing links it to `Protocol`, which typeshed declares as
`Protocol: _SpecialForm`. Typeshed cannot express it, which is why #1204 concluded a checker has to
hard-code it — and ty already does, for the bare form. The repro above is stdlib only, so there is no
stub to fix.

### Versions

- ty 0.0.72 (latest) and 0.0.71 — both affected
- pyright 1.1.411, mypy 2.3.1 — do not model the metaclass, so no diagnostic
- Python 3.12

**Comment by carljm** [comment]
I looked into fixing this, but it seems like we have a design decision to make here. If we apply this understanding of `Protocol` meta type generally, then we will understand `Mapping.register`, but the cost will be that we think some built-in types have `ABCMeta` metaclass (because typeshed lies about their real inheritance) when they don't at runtime, meaning we may emit false-positive bad-MRO errors; mypy documents this here: https://mypy.readthedocs.io/en/stable/metaclasses.html#gotchas-and-limitations-of-metaclass-support

Pyrefly takes a different route, which is that it only supports understanding the meta-type of protocols if they are defined in `.py` files, not in stub files (thus excluding typeshed): https://github.com/facebook/pyrefly/blob/630366a32e4f16b759689c093a4cf83d21a1038f/pyrefly/lib/alt/class/class_metadata.rs#L204-L210 -- this approach avoids the bad MRO false positives that mypy can have when subclassing some builtins, but it means we still would not understanding the existence of `Mapping.register` (and pyrefly does not), so arguably it would not even fix this issue as originally reported.

I'm inclined to follow mypy, not pyrefly, here, on the general principle that we have no choice but to trust typeshed, and where it lies, users have to deal with the consequences of those lies. Or put differently, that's typeshed's problem, not ours.
